packages/angular_devkit/architect/src/architect.ts

```diff
@@ -0,0 +1,262 @@
+import { resolve as nodeResolve } from '@angular-devkit/core/node';
+import { existsSync, readFileSync } from 'fs';
+import { dirname, resolve as pathResolve } from 'path';
+import { Observable } from 'rxjs/Observable';
+import { of } from 'rxjs/observable/of';
+import { _throw } from 'rxjs/observable/throw';
+import { concatMap } from 'rxjs/operators';
+import {
+  BuildEvent,
+  Builder,
+  BuilderClass,
+  BuilderContext,
+  BuilderDescription,
+  BuildersDescription,
+} from './builder';
+import { Workspace } from './workspace';
+
+
+export class ProjectNotFoundException extends BaseException {
+  constructor(name?: string) {
+    const nameOrDefault = name ? `Project '${name}'` : `Default project`;
+    super(`${nameOrDefault} could not be found in workspace.`);
+  }
+}
+
+export class TargetNotFoundException extends BaseException {
+  constructor(name?: string) {
+    const nameOrDefault = name ? `Target '${name}'` : `Default target`;
+    super(`${nameOrDefault} could not be found in workspace.`);
+  }
+}
+
+export class ConfigurationNotFoundException extends BaseException {
+  constructor(name: string) {
+    super(`Configuration '${name}' could not be found in project.`);
+  }
+}
+
+export class SchemaValidationException extends BaseException {
+  constructor(errors: string[]) {
+    super(`Schema validation failed with the following errors:\n  ${errors.join('\n  ')}`);
+  }
+}
+
+// TODO: break this exception apart into more granular ones.
+export class BuilderCannotBeResolvedException extends BaseException {
+  constructor(builder: string) {
+    super(`Builder '${builder}' cannot be resolved.`);
+  }
+}
+
+export class WorkspaceNotYetLoadedException extends BaseException {
+  constructor() { super(`Workspace needs to be loaded before Architect is used.`); }
+}
+
+export interface Target<OptionsT = {}> {
+  root: Path;
+  projectType: string;
+  builder: string;
+  options: OptionsT;
+}
+
+export class Architect {
+  private _workspaceSchema = pathResolve(__dirname, 'workspace-schema.json');
+  private _buildersSchema = pathResolve(__dirname, 'builders-schema.json');
+  private _workspace: Workspace;
+
+  constructor(private _root: Path, private _host: virtualFs.Host<{}>) { }
+
+  loadWorkspaceFromHost(workspacePath: Path) {
+    return this._host.read(join(this._root, workspacePath)).pipe(
+      concatMap((buffer) => {
+        const json = JSON.parse(virtualFs.fileBufferToString(buffer));
+
+        return this.loadWorkspaceFromJson(json);
+      }),
+    );
+  }
+
+  loadWorkspaceFromJson(json: Workspace) {
+    return this._validateAgainstSchema(json, this._workspaceSchema).pipe(
+      concatMap((validatedWorkspace: Workspace) => {
+        this._workspace = validatedWorkspace;
+
+        return of(this);
+      }),
+    );
+  }
+
+  getTarget<OptionsT>(
+    projectName?: string,
+    targetName?: string,
+    configurationName?: string,
+    additionalInput?: Partial<OptionsT>,
```

Comment:
Should `additionalInput` be first or have some overloads?  With the current order it doesn't appear possible to augment the default project/target without looking up the names and supplying them.

As an alternative, could `additionalInput` be dropped?  The target options could be adjusted after the call.

Reply:
I was actually thinking there should be an options object to avoid these problem. WDYT?